# Netflix Exploratory Data Analysis (EDA) using NumPy and Pandas

**Author:** Altamash

## Project Objective

This project performs complete Exploratory Data Analysis (EDA) on the Netflix  
Movies and TV Shows dataset using NumPy, Pandas, and Matplotlib.
### Goals

- Understand dataset structure.
- Clean missing and inconsistent data.
- Perform statistical analysis using NumPy.
- Analyze Netflix content using Pandas.
- Create business insights and visualizations.

In [291]:
# ==========================================
# Import Required Libraries
# ==========================================

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Notebook display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)

print("Libraries imported successfully.")

Libraries imported successfully.


## Load Netflix Dataset

The dataset is loaded from the data folder into a Pandas DataFrame.

In [292]:
# ==========================================
# Load Dataset
# ==========================================

df = pd.read_csv("../data/netflix_titles.csv")

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [293]:
# Dataset Shape

df.shape

(8807, 12)

In [294]:
#Display the first five records.
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [295]:
df.tail()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
8802,s8803,Movie,Zodiac,David Fincher,"Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...",United States,"November 20, 2019",2007,R,158 min,"Cult Movies, Dramas, Thrillers","A political cartoonist, a crime reporter and a..."
8803,s8804,TV Show,Zombie Dumb,NaN,NaN,NaN,"July 1, 2019",2018,TV-Y7,2 Seasons,"Kids' TV, Korean TV Shows, TV Comedies","While living alone in a spooky town, a young g..."
8804,s8805,Movie,Zombieland,Ruben Fleischer,"Jesse Eisenberg, Woody Harrelson, Emma Stone, ...",United States,"November 1, 2019",2009,R,88 min,"Comedies, Horror Movies",Looking to survive in a world taken over by zo...
8805,s8806,Movie,Zoom,Peter Hewitt,"Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...",United States,"January 11, 2020",2006,PG,88 min,"Children & Family Movies, Comedies","Dragged from civilian life, a former superhero..."
8806,s8807,Movie,Zubaan,Mozez Singh,"Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...",India,"March 2, 2019",2015,TV-14,111 min,"Dramas, International Movies, Music & Musicals",A scrappy but poor boy worms his way into a ty...


In [296]:
#List all feature names available in the dataset.

df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [297]:
#Count rows and columns separately.

print("numbers of rows : ",df.shape[0])
print("numbers of columns : ",df.shape[1])

numbers of rows :  8807
numbers of columns :  12


## Dataset Information

Understand data types, missing values, and memory usage.

In [298]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


## Statistical Summary

Generate summary statistics for numerical columns.

In [299]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


## Categorical Summary

Explore object-type columns.

In [300]:
df.describe(include="object")

,show_id,type,title,director,cast,country,date_added,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8803,8804,8807,8807
unique,8807,2,8807,4528,7692,748,1767,17,220,514,8775
top,s1,Movie,Dick Johnson Is Dead,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,1,19,19,2818,109,3207,1793,362,4


## Data Types

Check data type of each feature.

In [301]:
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

### Observation

date_added should be converted from object to datetime during data cleaning.

## Memory Usage

Check memory consumed by each column.

In [302]:
memory_usage = df.memory_usage(deep=True)

memory_usage

Index               132
show_id          474471
type             480930
title            593151
director         488185
cast            1422387
country          517757
date_added       561028
release_year      70456
rating           470514
duration         493486
listed_in        725748
description     1873527
dtype: int64

In [303]:
total_memory = memory_usage.sum() / (1024**2)

print(f"Total Memory Usage: {total_memory:.2f} MB")

Total Memory Usage: 7.79 MB


## Frequency Analysis

Count Movies and TV Shows.

In [304]:
df["type"].value_counts()

type
Movie      6131
TV Show    2676
Name: count, dtype: int64

In [305]:
df["type"].value_counts(normalize=True) * 100

type
Movie      69.615079
TV Show    30.384921
Name: proportion, dtype: float64

## Unique Value Analysis

In [306]:
df["country"].nunique()

748

In [307]:
df["country"].unique()[:10]

array(['United States', 'South Africa', nan, 'India',
       'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia',
       'United Kingdom', 'Germany, Czech Republic', 'Mexico', 'Turkey',
       'Australia'], dtype=object)

## EDA Summary Table

Create a compact summary containing data type, missing values, and unique values.

In [308]:
summary = pd.DataFrame({
    "Data Type": df.dtypes,
    "Missing Values": df.isnull().sum(),
    "Unique Values": df.nunique()
})

summary

,Data Type,Missing Values,Unique Values
show_id,object,0,8807
type,object,0,2
title,object,0,8807
director,object,2634,4528
cast,object,825,7692
country,object,831,748
date_added,object,10,1767
release_year,int64,0,74
rating,object,4,17
duration,object,3,220


## Initial Findings

### Key Observations

- The dataset contains 8,807 Netflix titles.
- There are 12 features describing each title.
- type contains two categories: Movies and TV Shows.
- date_added is stored as an object and requires datetime conversion.
- director, cast, and country contain missing values.
- Both numerical and categorical columns are present, requiring different analysis approaches.

# Missing Value Analysis

Before performing analysis, we need to identify missing values in every feature.
Missing values can affect statistics, visualizations, and machine learning models.

In [309]:
# Missing values in each column

df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [310]:
# Percentage of missing values

missing_percentage = (
    df.isnull().sum() / len(df)
) * 100

missing_percentage.sort_values(ascending=False)

director        29.908028
country          9.435676
cast             9.367549
date_added       0.113546
rating           0.045418
duration         0.034064
show_id          0.000000
type             0.000000
title            0.000000
release_year     0.000000
listed_in        0.000000
description      0.000000
dtype: float64

In [311]:
# Missing Value Report

missing_report = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": missing_percentage.round(2)
})

missing_report = missing_report.sort_values(
    by="Missing Percentage",
    ascending=False
)

missing_report

,Missing Count,Missing Percentage
director,2634,29.91
country,831,9.44
cast,825,9.37
date_added,10,0.11
rating,4,0.05
duration,3,0.03
show_id,0,0.00
type,0,0.00
title,0,0.00
release_year,0,0.00


In [312]:
# Fill missing values

df["director"] = df["director"].fillna("Unknown")

df["cast"] = df["cast"].fillna("Not Available")

df["country"] = df["country"].fillna("Unknown")

# Fill rating using mode
df["rating"] = df["rating"].fillna(df["rating"].mode()[0])

# Fill duration using mode
df["duration"] = df["duration"].fillna(df["duration"].mode()[0])

In [313]:
# Remove rows where date_added is missing

df = df.dropna(subset=["date_added"])

In [314]:
# Verify cleaning

df.isnull().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

In [315]:
df.shape

(8797, 12)

# Duplicate Analysis

Duplicate records can distort frequency analysis and business insights.

In [316]:
# Count duplicate rows

df.duplicated().sum()

0

# Rename Columns

Use consistent naming conventions for readability and production-quality code.

In [317]:
# Rename columns

df.rename(columns={
    "type": "content_type",
    "listed_in": "genre"
}, inplace=True)

In [318]:
df.columns

Index(['show_id', 'content_type', 'title', 'director', 'cast', 'country',
       'date_added', 'release_year', 'rating', 'duration', 'genre',
       'description'],
      dtype='object')

# Date Conversion

Convert the date_added column from string to datetime format.

In [319]:
type(df["date_added"][0])

str

In [320]:
df["date_added"][0]

'September 25, 2021'

In [321]:
# Remove leading/trailing spaces
df["date_added"] = df["date_added"].str.strip()

# Convert string to datetime
df["date_added"] = pd.to_datetime(
    df["date_added"],
    format="%B %d, %Y",
    errors="coerce"
)

In [322]:
df["date_added"].dtype

dtype('<M8[ns]')

# Feature Engineering from Dates

Extract useful temporal features from date_added.

In [323]:
# Create new features

df["year_added"] = df["date_added"].dt.year

df["month_added"] = df["date_added"].dt.month

df["month_name"] = df["date_added"].dt.month_name()

df["day_added"] = df["date_added"].dt.day

In [324]:
df[[
    "date_added",
    "year_added",
    "month_added",
    "month_name",
    "day_added"
]].head()

,date_added,year_added,month_added,month_name,day_added
0,2021-09-25,2021,9,September,25
1,2021-09-24,2021,9,September,24
2,2021-09-24,2021,9,September,24
3,2021-09-24,2021,9,September,24
4,2021-09-24,2021,9,September,24


In [325]:
df.sample(3)

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added
8672,s8673,Movie,Veerey Ki Wedding,Ashu Trikha,"Pulkit Samrat, Yuvika Chaudhry, Jimmy Shergill...",India,2019-12-31,2018,TV-14,130 min,"Comedies, Dramas, International Movies","Quick to throw punches in the name of justice,...",2019,12,December,31
4678,s4679,TV Show,Undercover Law,Unknown,"Viña Machado, Juana del Río, Luna Baxter, Vale...",Colombia,2018-08-31,2017,TV-MA,1 Season,"Crime TV Shows, International TV Shows, Spanis...",Female intelligence agents infiltrate the disp...,2018,8,August,31
1306,s1307,TV Show,Good Girls,Unknown,"Christina Hendricks, Retta, Mae Whitman, Matth...",United States,2021-02-16,2020,TV-MA,3 Seasons,"Crime TV Shows, TV Comedies, TV Dramas",Three suburban moms orchestrate a local grocer...,2021,2,February,16


# Remove Extra Spaces

Text columns often contain unwanted leading or trailing spaces.

In [326]:
# Remove spaces

text_columns = [
    "title",
    "director",
    "cast",
    "country",
    "rating",
    "duration",
    "genre"
]

for column in text_columns:
    df[column] = df[column].str.strip()

# Standardize Category Values

In [327]:
df["rating"].value_counts()

rating
TV-MA       3209
TV-14       2157
TV-PG        861
R            799
PG-13        490
TV-Y7        333
TV-Y         306
PG           287
TV-G         220
NR            79
G             41
TV-Y7-FV       6
NC-17          3
UR             3
74 min         1
84 min         1
66 min         1
Name: count, dtype: int64

# Clean Duration Column

Movies contain minutes.
TV Shows contain seasons.

We extract numerical duration separately.

In [328]:
# Extract numeric duration

df["duration_numeric"] = (
    df["duration"]
      .str.extract(r"(\d+)")
      .astype(int)
)

In [329]:
df[["duration", "duration_numeric"]].head()

,duration,duration_numeric
0,90 min,90
1,2 Seasons,2
2,1 Season,1
3,1 Season,1
4,2 Seasons,2


# Separate Movies and TV Shows

Create filtered datasets for independent analysis.

In [330]:
movies_df = df[df["content_type"] == "Movie"]

tvshows_df = df[df["content_type"] == "TV Show"]

In [331]:
print("Movies :", movies_df.shape)
print("TV Shows :", tvshows_df.shape)

Movies : (6131, 17)
TV Shows : (2666, 17)


# Validate Clean Dataset

Run final checks after cleaning.

In [332]:
print("Remaining Missing Values")
print(df.isnull().sum())

print("\nDuplicate Rows")
print(df.duplicated().sum())

print("\nDataset Shape")
print(df.shape)

Remaining Missing Values
show_id             0
content_type        0
title               0
director            0
cast                0
country             0
date_added          0
release_year        0
rating              0
duration            0
genre               0
description         0
year_added          0
month_added         0
month_name          0
day_added           0
duration_numeric    0
dtype: int64

Duplicate Rows
0

Dataset Shape
(8797, 17)


In [333]:
cleaning_summary = pd.DataFrame({
    "Metric": [
        "Rows",
        "Columns",
        "Duplicate Rows Removed",
        "Missing Values Remaining"
    ],
    "Value": [
        len(df),
        df.shape[1],
        0,
        df.isnull().sum().sum()
    ]
})

cleaning_summary

,Metric,Value
0,Rows,8797
1,Columns,17
2,Duplicate Rows Removed,0
3,Missing Values Remaining,0


## Cleaning Summary

### Completed Cleaning Steps

- Missing values handled using fill and drop strategies.
- Duplicate rows removed.
- Column names standardized.
- date_added converted to datetime.
- Extra spaces removed from text columns.
- Numerical duration extracted from text.
- New temporal features created (year_added, month_added, month_name, day_added).

The dataset is now ready for statistical analysis and visualization.

# Convert Pandas Columns to NumPy Arrays

NumPy performs mathematical operations on arrays faster than Python lists and is widely used in Machine Learning libraries.

In [334]:
# Convert columns into NumPy arrays

release_year_array = df["release_year"].to_numpy()
duration_array = df["duration_numeric"].to_numpy()

In [335]:
print(type(release_year_array))
print(type(duration_array))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


## Mean of Release Year

Mean gives the average release year of Netflix content.

In [336]:
mean_release_year = np.mean(release_year_array)
mean_duration = np.mean(duration_array)

print("Average Release Year :", round(mean_release_year, 2))
print("Average Duration :", round(mean_duration, 2))

Average Release Year : 2014.18
Average Duration : 69.9


## Median of Release Year

Median represents the middle value after sorting the data.

In [337]:
median_release_year = np.median(release_year_array)
median_duration = np.median(duration_array)

print("Median Release Year :", median_release_year)
print("Median Duration :", median_duration)

Median Release Year : 2017.0
Median Duration : 88.0


## Mode of Rating

Mode is the most frequently occurring value.

In [338]:
mode_rating = df["rating"].mode()[0]

print("Most Common Rating :", mode_rating)

Most Common Rating : TV-MA


## Minimum and Maximum Values

Find oldest and newest release years and shortest/longest durations.

In [339]:
print("Oldest Release Year :", np.min(release_year_array))
print("Newest Release Year :", np.max(release_year_array))

print("Shortest Duration :", np.min(duration_array))
print("Longest Duration :", np.max(duration_array))

Oldest Release Year : 1925
Newest Release Year : 2021
Shortest Duration : 1
Longest Duration : 312


## Range

Range measures spread between maximum and minimum values.

In [340]:
release_range = np.max(release_year_array) - np.min(release_year_array)
duration_range = np.max(duration_array) - np.min(duration_array)

print("Release Year Range :", release_range)
print("Duration Range :", duration_range)

Release Year Range : 96
Duration Range : 311


## Variance

Variance measures how far values are spread from the mean.

In [341]:
release_variance = np.var(release_year_array)
duration_variance = np.var(duration_array)

print("Variance (Release Year):", round(release_variance, 2))
print("Variance (Duration):", round(duration_variance, 2))

Variance (Release Year): 77.82
Variance (Duration): 2580.78


## Standard Deviation

Standard deviation is the square root of variance.

In [342]:
release_std = np.std(release_year_array)
duration_std = np.std(duration_array)

print("Standard Deviation (Release Year):", round(release_std, 2))
print("Standard Deviation (Duration):", round(duration_std, 2))

Standard Deviation (Release Year): 8.82
Standard Deviation (Duration): 50.8


## Percentile Analysis

Percentiles help understand data distribution.

In [343]:
percentiles = np.percentile(
    duration_array,
    [25, 50, 75, 90, 95]
)

print(percentiles)

[  2.  88. 106. 126. 139.]


In [344]:
#Quartiles individually

q1 = np.percentile(duration_array, 25)
q2 = np.percentile(duration_array, 50)
q3 = np.percentile(duration_array, 75)

print("Q1 :", q1)
print("Q2 :", q2)
print("Q3 :", q3)

Q1 : 2.0
Q2 : 88.0
Q3 : 106.0


## Interquartile Range (IQR)

IQR measures the spread of the middle 50% of the data.

In [345]:
iqr = q3 - q1

print("Interquartile Range :", iqr)

Interquartile Range : 104.0


## Outlier Detection Using IQR

In [346]:
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

outliers = df[
    (df["duration_numeric"] < lower_bound) |
    (df["duration_numeric"] > upper_bound)
]

outliers.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
717,s718,Movie,Headspace: Unwind Your Mind,Unknown,"Andy Puddicombe, Evelyn Lewis Prieto, Ginger D...",Unknown,2021-06-15,2021,TV-G,273 min,Documentaries,"Do you want to relax, meditate or sleep deeply...",2021,6,June,15,273
4253,s4254,Movie,Black Mirror: Bandersnatch,Unknown,"Fionn Whitehead, Will Poulter, Craig Parkinson...",United States,2018-12-28,2018,TV-MA,312 min,"Dramas, International Movies, Sci-Fi & Fantasy","In 1984, a young programmer begins to question...",2018,12,December,28,312


## NumPy Statistical Summary

In [347]:
statistics_summary = pd.DataFrame({
    "Metric": [
        "Mean",
        "Median",
        "Minimum",
        "Maximum",
        "Variance",
        "Standard Deviation"
    ],
    "Release Year": [
        np.mean(release_year_array),
        np.median(release_year_array),
        np.min(release_year_array),
        np.max(release_year_array),
        np.var(release_year_array),
        np.std(release_year_array)
    ],
    "Duration": [
        np.mean(duration_array),
        np.median(duration_array),
        np.min(duration_array),
        np.max(duration_array),
        np.var(duration_array),
        np.std(duration_array)
    ]
})

statistics_summary.round(2)

,Metric,Release Year,Duration
0,Mean,2014.18,69.90
1,Median,2017.00,88.00
2,Minimum,1925.00,1.00
3,Maximum,2021.00,312.00
4,Variance,77.82,2580.78
5,Standard Deviation,8.82,50.80


## Statistical Observations

### Key Findings

- Netflix content is primarily composed of relatively recent releases.
- Movie and TV Show durations vary significantly.
- The duration column contains outliers that can be analyzed separately.
- Median duration is more representative than mean because duration distribution is skewed.
- The IQR method identifies unusually long-duration content.

# Frequency Analysis using value_counts()

Identify the frequency of categorical values in the dataset.

In [348]:
# Count Movies and TV Shows

df["content_type"].value_counts()

content_type
Movie      6131
TV Show    2666
Name: count, dtype: int64

In [349]:
content_percentage = (
    df["content_type"]
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
)

content_percentage

content_type
Movie      69.69
TV Show    30.31
Name: proportion, dtype: float64

### Insight

Approximately 70% of Netflix content consists of Movies, while 30% consists of TV Shows.

# Most Common Ratings

Understand Netflix's content maturity distribution.

In [350]:
rating_counts = (
    df["rating"]
    .value_counts()
)

rating_counts

rating
TV-MA       3209
TV-14       2157
TV-PG        861
R            799
PG-13        490
TV-Y7        333
TV-Y         306
PG           287
TV-G         220
NR            79
G             41
TV-Y7-FV       6
NC-17          3
UR             3
74 min         1
84 min         1
66 min         1
Name: count, dtype: int64

In [351]:
rating_counts.head(10)

rating
TV-MA    3209
TV-14    2157
TV-PG     861
R         799
PG-13     490
TV-Y7     333
TV-Y      306
PG        287
TV-G      220
NR         79
Name: count, dtype: int64

# Unique Value Analysis

In [352]:
# Number of unique countries

df["country"].nunique()

749

In [353]:
# Display first 15 unique countries

df["country"].unique()[:15]

array(['United States', 'South Africa', 'Unknown', 'India',
       'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia',
       'United Kingdom', 'Germany, Czech Republic', 'Mexico', 'Turkey',
       'Australia', 'United States, India, France', 'Finland',
       'China, Canada, United States',
       'South Africa, United States, Japan', 'Nigeria'], dtype=object)

# Sorting Data

In [354]:
latest_content = df.sort_values(
    by="release_year",
    ascending=False
)

latest_content.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
693,s694,Movie,Ali & Ratu Ratu Queens,Lucky Kuswandi,"Iqbaal Ramadhan, Nirina Zubir, Asri Welas, Tik...",Unknown,2021-06-17,2021,TV-14,101 min,"Comedies, Dramas, International Movies","After his father's passing, a teenager sets ou...",2021,6,June,17,101
781,s782,Movie,Black Holes | The Edge of All We Know,Peter Galison,Not Available,Unknown,2021-06-02,2021,TV-14,99 min,Documentaries,Follow scientists on their quest to understand...,2021,6,June,2,99
762,s763,Movie,Sweet & Sour,Lee Kae-byeok,"Jang Ki-yong, Chae Soo-bin, Jung Soo-jung",South Korea,2021-06-04,2021,TV-14,103 min,"Comedies, International Movies, Romantic Movies",Faced with real-world opportunities and challe...,2021,6,June,4,103
763,s764,TV Show,Sweet Tooth,Unknown,"Nonso Anozie, Christian Convery, Adeel Akhtar,...",United States,2021-06-04,2021,TV-14,1 Season,"TV Action & Adventure, TV Dramas, TV Sci-Fi & ...",On a perilous adventure across a post-apocalyp...,2021,6,June,4,1
764,s765,Movie,Trippin' with the Kandasamys,Jayan Moodley,"Jailoshini Naidoo, Maeshni Naicker, Madhushan ...",South Africa,2021-06-04,2021,TV-14,94 min,"Comedies, International Movies, Romantic Movies","To rekindle their marriages, best friends-turn...",2021,6,June,4,94


In [355]:
oldest_content = df.sort_values(
    by="release_year"
)

oldest_content.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
4250,s4251,TV Show,Pioneers: First Women Filmmakers*,Unknown,Not Available,Unknown,2018-12-30,1925,TV-14,1 Season,TV Shows,This collection restores films from women who ...,2018,12,December,30,1
7790,s7791,Movie,Prelude to War,Frank Capra,Not Available,United States,2017-03-31,1942,TV-14,52 min,"Classic Movies, Documentaries",Frank Capra's documentary chronicles the rise ...,2017,3,March,31,52
8205,s8206,Movie,The Battle of Midway,John Ford,"Henry Fonda, Jane Darwell",United States,2017-03-31,1942,TV-14,18 min,"Classic Movies, Documentaries",Director John Ford captures combat footage of ...,2017,3,March,31,18
8739,s8740,Movie,Why We Fight: The Battle of Russia,"Frank Capra, Anatole Litvak",Not Available,United States,2017-03-31,1943,TV-PG,82 min,Documentaries,This installment of Frank Capra's acclaimed do...,2017,3,March,31,82
8763,s8764,Movie,WWII: Report from the Aleutians,John Huston,Not Available,United States,2017-03-31,1943,TV-PG,45 min,Documentaries,Filmmaker John Huston narrates this Oscar-nomi...,2017,3,March,31,45


In [356]:
df.sort_values(
    by=["release_year", "title"],
    ascending=[False, True]
).head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
1373,s1374,TV Show,50M2,Unknown,"Engin Öztürk, Kürşat Alnıaçık, Cengiz Bozkurt,...",Turkey,2021-01-27,2021,TV-MA,1 Season,"International TV Shows, TV Dramas, TV Thrillers","Seeking to uncover the truth about his past, a...",2021,1,January,27,1
1150,s1151,Movie,7 Yards: The Chris Norton Story,Jonathon Link,Chris Norton,United States,2021-03-30,2021,TV-PG,93 min,Documentaries,The true story of former college football play...,2021,3,March,30,93
850,s851,Movie,99 Songs,Vishwesh Krishnamoorthy,"Ehan Bhat, Edilsy Vargas, Manisha Koirala, Lis...",India,2021-05-21,2021,TV-14,131 min,"Dramas, International Movies, Music & Musicals",Challenged to compose 100 songs before he can ...,2021,5,May,21,131
851,s852,Movie,99 Songs (Tamil),Unknown,Not Available,Unknown,2021-05-21,2021,TV-14,131 min,"Dramas, International Movies, Music & Musicals",Challenged to compose 100 songs before he can ...,2021,5,May,21,131
852,s853,Movie,99 Songs (Telugu),Unknown,Not Available,Unknown,2021-05-21,2021,TV-14,131 min,"Dramas, International Movies, Music & Musicals",Challenged to compose 100 songs before he can ...,2021,5,May,21,131


# loc vs iloc

Select rows and columns using labels and positions.

In [357]:
df.loc[0]

show_id                                                            s1
content_type                                                    Movie
title                                            Dick Johnson Is Dead
director                                              Kirsten Johnson
cast                                                    Not Available
country                                                 United States
date_added                                        2021-09-25 00:00:00
release_year                                                     2020
rating                                                          PG-13
duration                                                       90 min
genre                                                   Documentaries
description         As her father nears the end of his life, filmm...
year_added                                                       2021
month_added                                                         9
month_name          

In [358]:
#Select Specific Columns

df.loc[
    0:4,
    ["title", "country", "release_year"]
]

,title,country,release_year
0,Dick Johnson Is Dead,United States,2020
1,Blood & Water,South Africa,2021
2,Ganglands,Unknown,2021
3,Jailbirds New Orleans,Unknown,2021
4,Kota Factory,India,2021


In [359]:
#Filter Using loc
df.loc[
    df["release_year"] >= 2020,
    ["title", "release_year", "country"]
].head()

,title,release_year,country
0,Dick Johnson Is Dead,2020,United States
1,Blood & Water,2021,South Africa
2,Ganglands,2021,Unknown
3,Jailbirds New Orleans,2021,Unknown
4,Kota Factory,2021,India


In [360]:
#iloc — Position Based

df.iloc[0]

show_id                                                            s1
content_type                                                    Movie
title                                            Dick Johnson Is Dead
director                                              Kirsten Johnson
cast                                                    Not Available
country                                                 United States
date_added                                        2021-09-25 00:00:00
release_year                                                     2020
rating                                                          PG-13
duration                                                       90 min
genre                                                   Documentaries
description         As her father nears the end of his life, filmm...
year_added                                                       2021
month_added                                                         9
month_name          

In [361]:
#Rows and Columns by Position

df.iloc[
    0:5,
    1:5
]

,content_type,title,director,cast
0,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available
1,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban..."
2,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi..."
3,TV Show,Jailbirds New Orleans,Unknown,Not Available
4,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K..."


# Filtering Rows

In [362]:
#Movies Only

movies = df[
    df["content_type"] == "Movie"
]

movies.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021,9,September,25,90
6,s7,Movie,My Little Pony: A New Generation,"Robert Cullen, José Luis Ucha","Vanessa Hudgens, Kimiko Glenn, James Marsden, ...",Unknown,2021-09-24,2021,PG,91 min,Children & Family Movies,Equestria's divided. But a bright-eyed hero be...,2021,9,September,24,91
7,s8,Movie,Sankofa,Haile Gerima,"Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...","United States, Ghana, Burkina Faso, United Kin...",2021-09-24,1993,TV-MA,125 min,"Dramas, Independent Movies, International Movies","On a photo shoot in Ghana, an American model s...",2021,9,September,24,125
9,s10,Movie,The Starling,Theodore Melfi,"Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...",United States,2021-09-24,2021,PG-13,104 min,"Comedies, Dramas",A woman adjusting to life after a loss contend...,2021,9,September,24,104
12,s13,Movie,Je Suis Karl,Christian Schwochow,"Luna Wedler, Jannis Niewöhner, Milan Peschel, ...","Germany, Czech Republic",2021-09-23,2021,TV-MA,127 min,"Dramas, International Movies",After most of her family is murdered in a terr...,2021,9,September,23,127


In [363]:
#TV Shows Only

tv_shows = df[
    df["content_type"] == "TV Show"
]

tv_shows.head()


,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
1,s2,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",2021,9,September,24,2
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,2021,9,September,24,1
3,s4,TV Show,Jailbirds New Orleans,Unknown,Not Available,Unknown,2021-09-24,2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo...",2021,9,September,24,1
4,s5,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,2021,9,September,24,2
5,s6,TV Show,Midnight Mass,Mike Flanagan,"Kate Siegel, Zach Gilford, Hamish Linklater, H...",Unknown,2021-09-24,2021,TV-MA,1 Season,"TV Dramas, TV Horror, TV Mysteries",The arrival of a charismatic young priest brin...,2021,9,September,24,1


In [364]:
#Movies Released After 2020

recent_movies = df[
    (df["content_type"] == "Movie") &
    (df["release_year"] >= 2020)
]

recent_movies.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021,9,September,25,90
6,s7,Movie,My Little Pony: A New Generation,"Robert Cullen, José Luis Ucha","Vanessa Hudgens, Kimiko Glenn, James Marsden, ...",Unknown,2021-09-24,2021,PG,91 min,Children & Family Movies,Equestria's divided. But a bright-eyed hero be...,2021,9,September,24,91
9,s10,Movie,The Starling,Theodore Melfi,"Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...",United States,2021-09-24,2021,PG-13,104 min,"Comedies, Dramas",A woman adjusting to life after a loss contend...,2021,9,September,24,104
12,s13,Movie,Je Suis Karl,Christian Schwochow,"Luna Wedler, Jannis Niewöhner, Milan Peschel, ...","Germany, Czech Republic",2021-09-23,2021,TV-MA,127 min,"Dramas, International Movies",After most of her family is murdered in a terr...,2021,9,September,23,127
13,s14,Movie,Confessions of an Invisible Girl,Bruno Garotti,"Klara Castanho, Lucca Picon, Júlia Gomes, Marc...",Unknown,2021-09-22,2021,TV-PG,91 min,"Children & Family Movies, Comedies",When the clever but socially-awkward Tetê join...,2021,9,September,22,91


In [365]:
#OR Condition (|)

india_or_usa = df[
    (df["country"] == "India") |
    (df["country"] == "United States")
]

india_or_usa.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021,9,September,25,90
4,s5,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,2021,9,September,24,2
9,s10,Movie,The Starling,Theodore Melfi,"Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...",United States,2021-09-24,2021,PG-13,104 min,"Comedies, Dramas",A woman adjusting to life after a loss contend...,2021,9,September,24,104
15,s16,TV Show,Dear White People,Unknown,"Logan Browning, Brandon P. Bell, DeRon Horton,...",United States,2021-09-22,2021,TV-MA,4 Seasons,"TV Comedies, TV Dramas",Students of color navigate the daily slights a...,2021,9,September,22,4
24,s25,Movie,Jeans,S. Shankar,"Prashanth, Aishwarya Rai Bachchan, Sri Lakshmi...",India,2021-09-21,1998,TV-14,166 min,"Comedies, International Movies, Romantic Movies",When the father of the man she loves insists t...,2021,9,September,21,166


In [366]:
#Using isin()
#Multiple Countries

countries = ["India", "United States", "United Kingdom"]

df[
    df["country"].isin(countries)
].head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021,9,September,25,90
4,s5,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,2021,9,September,24,2
8,s9,TV Show,The Great British Baking Show,Andy Devonshire,"Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...",United Kingdom,2021-09-24,2021,TV-14,9 Seasons,"British TV Shows, Reality TV",A talented batch of amateur bakers face off in...,2021,9,September,24,9
9,s10,Movie,The Starling,Theodore Melfi,"Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...",United States,2021-09-24,2021,PG-13,104 min,"Comedies, Dramas",A woman adjusting to life after a loss contend...,2021,9,September,24,104
15,s16,TV Show,Dear White People,Unknown,"Logan Browning, Brandon P. Bell, DeRon Horton,...",United States,2021-09-22,2021,TV-MA,4 Seasons,"TV Comedies, TV Dramas",Students of color navigate the daily slights a...,2021,9,September,22,4


In [367]:
#NOT Condition (~)
#Non-Indian Content

non_indian = df[
    ~df["country"].isin(["India"])
]

non_indian.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Not Available,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021,9,September,25,90
1,s2,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",2021,9,September,24,2
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,2021,9,September,24,1
3,s4,TV Show,Jailbirds New Orleans,Unknown,Not Available,Unknown,2021-09-24,2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo...",2021,9,September,24,1
5,s6,TV Show,Midnight Mass,Mike Flanagan,"Kate Siegel, Zach Gilford, Hamish Linklater, H...",Unknown,2021-09-24,2021,TV-MA,1 Season,"TV Dramas, TV Horror, TV Mysteries",The arrival of a charismatic young priest brin...,2021,9,September,24,1


# GroupBy Analysis

Group data by categories to generate business insights.

In [368]:
#Movies vs TV Shows Count

df.groupby("content_type").count()["show_id"]


content_type
Movie      6131
TV Show    2666
Name: show_id, dtype: int64

In [369]:
#Content Added Each Year

yearly_content = (
    df.groupby("year_added")
    .size()
    .sort_values(ascending=False)
)

yearly_content.head(10)


year_added
2019    2016
2020    1879
2018    1649
2021    1498
2017    1188
2016     429
2015      82
2014      24
2011      13
2013      11
dtype: int64

In [370]:
#Top Countries

top_countries = (
    df.groupby("country")
      .size()
      .sort_values(ascending=False)
      .head(10)
)

top_countries

country
United States     2812
India              972
Unknown            830
United Kingdom     418
Japan              244
South Korea        199
Canada             181
Spain              145
France             124
Mexico             110
dtype: int64

In [371]:
#Top Genres

top_genres = (
    df.groupby("genre")
      .size()
      .sort_values(ascending=False)
      .head(10)
)

top_genres

genre
Dramas, International Movies                        362
Documentaries                                       359
Stand-Up Comedy                                     334
Comedies, Dramas, International Movies              274
Dramas, Independent Movies, International Movies    252
Kids' TV                                            219
Children & Family Movies                            215
Children & Family Movies, Comedies                  201
Documentaries, International Movies                 186
Dramas, International Movies, Romantic Movies       180
dtype: int64

In [372]:
#Average Release Year by Content Type

df.groupby("content_type")["release_year"].mean()

content_type
Movie      2013.121514
TV Show    2016.625656
Name: release_year, dtype: float64

# Aggregation with agg()

Calculate multiple statistics simultaneously.

In [373]:
year_statistics = (
    df.groupby("year_added")
      .agg(
          total_titles=("show_id", "count"),
          average_release_year=("release_year", "mean"),
          oldest_release=("release_year", "min"),
          newest_release=("release_year", "max")
      )
)

year_statistics.head()

,total_titles,average_release_year,oldest_release,newest_release
year_added,,,,
2008,2,2006.500000,2006,2007
2009,2,2008.000000,2007,2009
2010,1,1987.000000,1987,1987
2011,13,1991.076923,1986,2010
2012,3,2011.000000,2010,2012


# String Operations

In [374]:
#Find Movies Containing "Love"

love_movies = df[
    df["title"].str.contains("Love", case=False, na=False)
]

love_movies.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
25,s26,TV Show,Love on the Spectrum,Unknown,Brooke Satchwell,Australia,2021-09-21,2021,TV-14,2 Seasons,"Docuseries, International TV Shows, Reality TV",Finding love can be hard for anyone. For young...,2021,9,September,21,2
158,s159,Movie,Love Don't Cost a Thing,Troy Byer,"Nick Cannon, Christina Milian, Kenan Thompson,...",United States,2021-09-01,2003,PG-13,101 min,"Comedies, Romantic Movies",A nerdy teen tries to make himself cool by ass...,2021,9,September,1,101
159,s160,Movie,Love in a Puff,Pang Ho-cheung,"Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...",Hong Kong,2021-09-01,2010,TV-MA,103 min,"Comedies, Dramas, International Movies",When the Hong Kong government enacts a ban on ...,2021,9,September,1,103
206,s207,Movie,"LSD: Love, Sex Aur Dhokha",Dibakar Banerjee,"Nushrat Bharucha, Anshuman Jha, Neha Chauhan, ...",India,2021-08-27,2010,TV-MA,112 min,"Dramas, Independent Movies, International Movies",This provocative drama examines how the voyeur...,2021,8,August,27,112
227,s228,Movie,Really Love,Angel Kristi Williams,"Kofi Siriboe, Yootha Wong-Loi-Sing, Michael Ea...",United States,2021-08-25,2020,TV-MA,95 min,"Dramas, Independent Movies, Romantic Movies",A rising Black painter tries to break into a c...,2021,8,August,25,95


In [375]:
#Count Horror Content

horror_titles = df[
    df["genre"].str.contains("Horror", case=False, na=False)
]

len(horror_titles)

432

# Country-wise Analysis

In [376]:
#Indian Content

india_content = df[
    df["country"].str.contains("India", na=False)
]

india_content.head()

,show_id,content_type,title,director,cast,country,date_added,release_year,rating,duration,genre,description,year_added,month_added,month_name,day_added,duration_numeric
4,s5,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,2021,9,September,24,2
24,s25,Movie,Jeans,S. Shankar,"Prashanth, Aishwarya Rai Bachchan, Sri Lakshmi...",India,2021-09-21,1998,TV-14,166 min,"Comedies, International Movies, Romantic Movies",When the father of the man she loves insists t...,2021,9,September,21,166
29,s30,Movie,Paranoia,Robert Luketic,"Liam Hemsworth, Gary Oldman, Amber Heard, Harr...","United States, India, France",2021-09-19,2013,PG-13,106 min,Thrillers,"Blackmailed by his company's CEO, a low-level ...",2021,9,September,19,106
39,s40,TV Show,Chhota Bheem,Unknown,"Vatsal Dubey, Julie Tejwani, Rupa Bhimani, Jig...",India,2021-09-16,2021,TV-Y7,3 Seasons,Kids' TV,"A brave, energetic little boy with superhuman ...",2021,9,September,16,3
50,s51,TV Show,Dharmakshetra,Unknown,"Kashmira Irani, Chandan Anand, Dinesh Mehta, A...",India,2021-09-15,2014,TV-PG,1 Season,"International TV Shows, TV Dramas, TV Sci-Fi &...","After the ancient Great War, the god Chitragup...",2021,9,September,15,1


In [377]:
# Count Indian Movies vs TV Shows

india_content["content_type"].value_counts()

content_type
Movie      962
TV Show     84
Name: count, dtype: int64

In [378]:
#USA Content

usa_content = df[
    df["country"].str.contains("United States", na=False)
]

usa_content["content_type"].value_counts()

content_type
Movie      2752
TV Show     932
Name: count, dtype: int64

# Business Summary Tables

In [379]:
country_analysis = (
    df.groupby("country")
      .size()
      .reset_index(name="Total Content")
      .sort_values(
          by="Total Content",
          ascending=False
      )
)

country_analysis.head(10)

,country,Total Content
603,United States,2812
251,India,972
736,Unknown,830
506,United Kingdom,418
319,Japan,244
435,South Korea,199
58,Canada,181
444,Spain,145
162,France,124
342,Mexico,110


In [380]:
#Top Years Table

year_analysis = (
    df.groupby("year_added")
      .size()
      .reset_index(name="Titles Added")
      .sort_values(
          by="Titles Added",
          ascending=False
      )
)

year_analysis.head()

,year_added,Titles Added
11,2019,2016
12,2020,1879
10,2018,1649
13,2021,1498
9,2017,1188


In [381]:
#Save for Outputs Folder

country_analysis.to_csv(
    "../outputs/country_analysis.csv",
    index=False
)

year_analysis.to_csv(
    "../outputs/yearly_content.csv",
    index=False
)

# Pandas EDA Observations

### Key Findings

- Movies dominate Netflix's catalog.
- Content additions peaked during specific years.
- The United States contributes the largest amount of content.
- India is among the top content-producing countries.
- TV-MA is the most frequent maturity rating.
- Drama and International Movies are among the most common genres.